File: clients/kotlin/ContractTest.kt
// Runs the shared SDK contract scenario against real compiled binaries.
//
// Every SESAME SDK drives this same sequence, so a divergence in any one
// client fails its own suite rather than hiding behind a mock. The harness is
// plain Kotlin: adding a test framework would contradict the shim's
// dependency-free promise.

package ma.del.sesame

import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import kotlin.system.exitProcess

private var checks = 0

private fun areEqual(expected: Any?, actual: Any?, what: String) {
    checks++
    if (expected?.toString() != actual?.toString()) {
        System.err.println("$what: expected $expected, got $actual")
        exitProcess(1)
    }
}

private fun field(value: Any?, key: String): Any? {
    val map = value as? Map<*, *> ?: run {
        System.err.println("expected an object, got $value")
        exitProcess(1)
    }
    return map[key]
}

private fun codeOf(action: () -> Unit): String {
    try {
        action()
    } catch (error: ProtocolException) {
        return error.code
    }
    System.err.println("expected a protocol error")
    exitProcess(1)
}

private fun repositoryRoot(): Path {
    var candidate: Path? = Path.of("").toAbsolutePath()
    while (candidate != null) {
        if (Files.isRegularFile(candidate.resolve("go.mod"))) return candidate
        candidate = candidate.parent
    }
    error("could not locate the repository root")
}

private fun build(root: Path, workspace: Path, name: String, source: String): Path {
    val output = workspace.resolve(name)
    val builder = ProcessBuilder("go", "build", "-trimpath", "-o", output.toString(), source)
        .directory(root.toFile())
        .inheritIO()
    builder.environment()["CGO_ENABLED"] = "0"
    builder.environment()["GOTOOLCHAIN"] = "auto"
    check(builder.start().waitFor() == 0) { "go build $source failed" }
    return output
}

private fun initializeDeployment(
    root: Path,
    workspace: Path,
    sesameBinary: Path,
    fakeFylo: Path,
): Path {
    val deployment = workspace.resolve("deployment")
    val builder = ProcessBuilder(
        sesameBinary.toString(),
        "init",
        "--deployment",
        deployment.toString(),
        "--fylo-binary",
        fakeFylo.toString(),
        "--issuer",
        "https://id.example",
    )
        .directory(root.toFile())
        .redirectOutput(ProcessBuilder.Redirect.DISCARD)
        .redirectError(ProcessBuilder.Redirect.INHERIT)
    check(builder.start().waitFor() == 0) { "sesame init failed" }
    return deployment
}

fun main() {
    val root = repositoryRoot()
    val workspace = Files.createTempDirectory("sesame-kotlin-contract-")
    try {
        val sesameBinary = build(root, workspace, "sesame", "./cmd/sesame")
        val fakeFylo = build(
            root, workspace, "fake-fylo", "./internal/adapters/fylo/testdata/fakefylo",
        )
        val deployment = initializeDeployment(root, workspace, sesameBinary, fakeFylo)

        Client(
            Options(
                binary = sesameBinary.toString(),
                deployment = deployment.toString(),
            ),
        ).use { client ->
            // System operations report a storage-backed process.
            areEqual("ok", field(client.ping(), "status"), "ping status")
            areEqual("ok", field(client.readiness(), "status"), "readiness status")
            areEqual("sesame", field(client.version(), "name"), "version name")
            areEqual(true, field(client.metrics(), "storage_configured"), "storage configured")

            // Tenant and principal.
            val tenantId = field(field(client.tenantBootstrap("acme"), "tenant"), "tenant_id")
                as String
            val principal = client.principalCreate(tenantId, "human", "email", "Alice@Example.com")
            val principalId = field(principal, "principal_id") as String
            areEqual(
                "alice@example.com",
                field(field(principal, "identifier"), "value"),
                "the engine normalizes the identifier",
            )

            // A duplicate claim returns the stable conflict code.
            areEqual(
                "identifier_conflict",
                codeOf {
                    client.principalCreate(tenantId, "workload", "email", "alice@example.com")
                },
                "duplicate identifier",
            )

            // Authorization: deny by default, allow after a grant, deny after revoke.
            val roleId = field(
                client.roleCreate(
                    tenantId,
                    "reader",
                    listOf(mapOf("action" to "doc:read", "resource" to "project:*")),
                ),
                "role_id",
            ) as String

            val denied = client.decide(tenantId, principalId, "doc:read", "project:alpha")
            areEqual("deny", field(denied, "decision"), "pre-grant decision")
            areEqual("deny_no_grant", field(denied, "reason_code"), "pre-grant reason")

            val grantId = field(
                client.grantCreate(tenantId, principalId, roleId), "grant_id",
            ) as String
            val allowed = client.decide(tenantId, principalId, "doc:read", "project:alpha")
            areEqual("allow", field(allowed, "decision"), "post-grant decision")
            areEqual("allow_role_grant", field(allowed, "reason_code"), "post-grant reason")

            client.grantRevoke(grantId)
            areEqual(
                "deny",
                field(client.decide(tenantId, principalId, "doc:read", "project:alpha"), "decision"),
                "post-revoke decision",
            )

            // Authentication: password login, session verification, revocation.
            val password = "correct horse battery staple"
            client.setPassword(principalId, password)
            val begun = client.authnBegin(tenantId, "email", "Alice@Example.com")
            areEqual("awaiting_factor", field(begun, "state"), "begin state")
            val transactionId = field(begun, "transaction_id") as String

            val wrong = client.authnVerifyPassword(transactionId, "wrong password value")
            areEqual(4L, field(wrong, "attempts_left"), "attempts after a wrong password")

            val verified = client.authnVerifyPassword(transactionId, password)
            areEqual("password", field(verified, "assurance"), "assurance")

            val issued = client.authnComplete(transactionId, 3600)
            val sessionId = field(issued, "session_id") as String
            val secret = field(issued, "session_secret") as String

            val session = client.sessionVerify(sessionId, secret)
            areEqual(principalId, field(session, "principal_id"), "verified session principal")
            areEqual(
                null,
                field(session, "secret_digest"),
                "the stored digest must not cross the boundary",
            )

            areEqual(
                "session_not_found",
                codeOf { client.sessionVerify(sessionId, "nope") },
                "wrong session secret",
            )
            client.sessionRevoke(sessionId, "test")
            areEqual(
                "session_inactive",
                codeOf { client.sessionVerify(sessionId, secret) },
                "revoked session",
            )

            // An unknown identifier is indistinguishable from a known one.
            val unknown = client.authnBegin(tenantId, "email", "ghost@example.com")
            areEqual(field(begun, "state"), field(unknown, "state"), "unknown identifier state")
            val attempt = client.authnVerifyPassword(
                field(unknown, "transaction_id") as String, password,
            )
            areEqual(4L, field(attempt, "attempts_left"), "unknown identifier attempts")
            areEqual(null, field(attempt, "assurance"), "unknown identifier assurance")

            // Unknown records return stable codes rather than transport failures.
            areEqual(
                "tenant_not_found",
                codeOf { client.tenantGetByName("missing") },
                "missing tenant",
            )
            areEqual(
                "operation_not_found",
                codeOf { client.request("identity.unknown") },
                "unknown operation",
            )

            standardsDispatchScenario(client)
        }
        println("ok\tkotlin contract scenario\t$checks checks")
    } finally {
        workspace.toFile().walkBottomUp().forEach(File::delete)
    }
}

private fun standardsDispatchScenario(client: Client) {
    val endpoints = mapOf(
        "authorization_endpoint" to "/authorize",
        "token_endpoint" to "/token",
        "jwks_uri" to "/.well-known/jwks.json",
        "introspection_endpoint" to "/introspect",
        "revocation_endpoint" to "/revoke",
        "end_session_endpoint" to "/logout",
    )
    val discovery = client.standardsDispatch(
        mapOf(
            "contract_version" to "unsupported",
            "endpoint" to "oidc.discovery",
            "method" to "GET",
            "endpoints" to endpoints,
        ),
    )
    areEqual("1", field(discovery, "contract_version"), "standards contract version")
    areEqual(200L, field(discovery, "status"), "standards discovery status")
    areEqual(
        "application/json",
        field(field(discovery, "headers"), "content-type"),
        "standards discovery content type",
    )
    areEqual(
        "https://id.example",
        field(field(discovery, "body"), "issuer"),
        "standards discovery issuer",
    )
    areEqual(
        "https://id.example/token",
        field(field(discovery, "body"), "token_endpoint"),
        "standards discovery token endpoint",
    )

    areEqual(
        "invalid_request",
        codeOf {
            client.standardsDispatch(
                mapOf("endpoint" to "oidc.userinfo", "method" to "GET"),
            )
        },
        "unknown standards endpoint",
    )
}

File: clients/rust/contract_test.rs
//! Runs the shared SDK contract scenario against real compiled binaries.
//!
//! Every SESAME SDK drives this same sequence, so a divergence in any one
//! client fails its own suite rather than hiding behind a mock.

#[path = "sesame.rs"]
mod sesame;

use sesame::{Client, Error, Options, Value};
use std::collections::BTreeMap;
use std::env;
use std::fs;
use std::path::PathBuf;
use std::process::Command;

/// Finds the repository root by walking up from the working directory until
/// go.mod appears, so the test runs the same under cargo and plain rustc.
fn repository_root() -> PathBuf {
    let mut candidate = env::current_dir().expect("current directory");
    loop {
        if candidate.join("go.mod").is_file() {
            return candidate;
        }
        if !candidate.pop() {
            panic!("could not locate the repository root from the working directory");
        }
    }
}

fn build(root: &PathBuf, workspace: &PathBuf, name: &str, source: &str) -> PathBuf {
    let output = workspace.join(name);
    let status = Command::new("go")
        .args(["build", "-trimpath", "-o"])
        .arg(&output)
        .arg(source)
        .current_dir(root)
        .env("CGO_ENABLED", "0")
        .env("GOTOOLCHAIN", "auto")
        .status()
        .expect("run go build");
    assert!(status.success(), "go build {source} failed");
    output
}

fn expect_str(value: &Value, key: &str) -> String {
    value
        .get(key)
        .and_then(Value::as_str)
        .unwrap_or_else(|| panic!("missing string field {key} in {value:?}"))
        .to_string()
}

#[test]
fn sdk_contract_scenario() {
    let root = repository_root();
    let workspace = env::temp_dir().join(format!("sesame-rust-contract-{}", std::process::id()));
    fs::create_dir_all(&workspace).expect("create workspace");

    let sesame_binary = build(&root, &workspace, "sesame", "./cmd/sesame");
    let fake_fylo = build(
        &root,
        &workspace,
        "fake-fylo",
        "./internal/adapters/fylo/testdata/fakefylo",
    );
    let fylo_root = workspace.join("root");
    fs::create_dir_all(&fylo_root).expect("create root");

    let mut client = Client::start(Options {
        binary: sesame_binary.to_string_lossy().into_owned(),
        fylo_binary: fake_fylo.to_string_lossy().into_owned(),
        fylo_root: fylo_root.to_string_lossy().into_owned(),
        ..Options::default()
    })
    .expect("start sesame");

    // System operations report a storage-backed process.
    assert_eq!(
        client.ping().expect("ping").get("status").and_then(Value::as_str),
        Some("ok")
    );
    assert_eq!(
        client
            .readiness()
            .expect("readiness")
            .get("status")
            .and_then(Value::as_str),
        Some("ok")
    );
    assert_eq!(
        client
            .version()
            .expect("version")
            .get("name")
            .and_then(Value::as_str),
        Some("sesame")
    );
    assert_eq!(
        client
            .metrics()
            .expect("metrics")
            .get("storage_configured")
            .and_then(Value::as_bool),
        Some(true)
    );

    // Standards dispatch pins contract v1 and rejects boundary injection.
    let standards = client
        .standards_dispatch(sesame::object(&[
            ("contract_version", "unsupported"),
            ("endpoint", "oidc.token"),
            ("method", "GET"),
        ]))
        .expect("standards.dispatch method response");
    assert_eq!(
        standards.get("contract_version").and_then(Value::as_str),
        Some("1")
    );
    assert_eq!(standards.get("status").and_then(Value::as_f64), Some(405.0));
    assert_eq!(
        standards
            .get("headers")
            .and_then(|headers| headers.get("allow"))
            .and_then(Value::as_str),
        Some("POST")
    );
    assert_eq!(
        standards
            .get("headers")
            .and_then(|headers| headers.get("content-type"))
            .and_then(Value::as_str),
        Some("application/json")
    );
    assert_eq!(
        standards
            .get("body")
            .and_then(|body| body.get("error"))
            .and_then(Value::as_str),
        Some("invalid_request")
    );

    let rejected = client
        .standards_dispatch(sesame::object(&[
            ("contract_version", "unsupported"),
            ("endpoint", "oidc.token"),
            ("method", "POST"),
            ("authorization", "Basic safe\r\nX-Injected: yes"),
        ]))
        .expect_err("control-character authorization");
    assert_eq!(rejected.code(), Some("invalid_request"));

    // Tenant and principal.
    let tenant = client.tenant_bootstrap("acme").expect("bootstrap");
    let tenant_id = expect_str(tenant.get("tenant").expect("tenant"), "tenant_id");
    let principal = client
        .principal_create(&tenant_id, "human", "email", "Alice@Example.com")
        .expect("principal.create");
    let principal_id = expect_str(&principal, "principal_id");
    assert_eq!(
        principal
            .get("identifier")
            .and_then(|identifier| identifier.get("value"))
            .and_then(Value::as_str),
        Some("alice@example.com"),
        "the engine normalizes the identifier"
    );

    // A duplicate claim returns the stable conflict code.
    let conflict = client
        .principal_create(&tenant_id, "workload", "email", "alice@example.com")
        .expect_err("duplicate identifier");
    assert_eq!(conflict.code(), Some("identifier_conflict"));

    // Authorization: deny by default, allow after a grant, deny after revoke.
    let mut permission = BTreeMap::new();
    permission.insert("action".to_string(), Value::String("doc:read".into()));
    permission.insert("resource".to_string(), Value::String("project:*".into()));
    let role = client
        .role_create(&tenant_id, "reader", vec![Value::Object(permission)])
        .expect("role.create");
    let role_id = expect_str(&role, "role_id");

    let denied = client
        .decide(&tenant_id, &principal_id, "doc:read", "project:alpha")
        .expect("decide");
    assert_eq!(denied.get("decision").and_then(Value::as_str), Some("deny"));
    assert_eq!(
        denied.get("reason_code").and_then(Value::as_str),
        Some("deny_no_grant")
    );

    let grant = client
        .grant_create(&tenant_id, &principal_id, &role_id)
        .expect("grant.create");
    let grant_id = expect_str(&grant, "grant_id");
    let allowed = client
        .decide(&tenant_id, &principal_id, "doc:read", "project:alpha")
        .expect("decide");
    assert_eq!(allowed.get("decision").and_then(Value::as_str), Some("allow"));
    assert_eq!(
        allowed.get("reason_code").and_then(Value::as_str),
        Some("allow_role_grant")
    );

    client.grant_revoke(&grant_id).expect("grant.revoke");
    let after_revoke = client
        .decide(&tenant_id, &principal_id, "doc:read", "project:alpha")
        .expect("decide");
    assert_eq!(
        after_revoke.get("decision").and_then(Value::as_str),
        Some("deny")
    );

    // Authentication: password login, session verification, revocation.
    let password = "correct horse battery staple";
    client
        .set_password(&principal_id, password)
        .expect("set_password");
    let begun = client
        .authn_begin(&tenant_id, "email", "Alice@Example.com")
        .expect("authn.begin");
    assert_eq!(
        begun.get("state").and_then(Value::as_str),
        Some("awaiting_factor")
    );
    let transaction_id = expect_str(&begun, "transaction_id");

    let wrong = client
        .authn_verify_password(&transaction_id, "wrong password value")
        .expect("wrong password");
    assert_eq!(wrong.get("attempts_left").and_then(Value::as_f64), Some(4.0));

    let verified = client
        .authn_verify_password(&transaction_id, password)
        .expect("verify");
    assert_eq!(
        verified.get("assurance").and_then(Value::as_str),
        Some("password")
    );

    let issued = client
        .authn_complete(&transaction_id, 3600)
        .expect("complete");
    let session_id = expect_str(&issued, "session_id");
    let secret = expect_str(&issued, "session_secret");

    let session = client
        .session_verify(&session_id, &secret)
        .expect("session.verify");
    assert_eq!(
        session.get("principal_id").and_then(Value::as_str),
        Some(principal_id.as_str())
    );
    assert!(
        session.get("secret_digest").is_none(),
        "the stored digest must not cross the boundary"
    );

    let wrong_secret = client
        .session_verify(&session_id, "nope")
        .expect_err("wrong secret");
    assert_eq!(wrong_secret.code(), Some("session_not_found"));

    client
        .session_revoke(&session_id, "test")
        .expect("session.revoke");
    let revoked = client
        .session_verify(&session_id, &secret)
        .expect_err("revoked session");
    assert_eq!(revoked.code(), Some("session_inactive"));

    // An unknown identifier is indistinguishable from a known one.
    let unknown = client
        .authn_begin(&tenant_id, "email", "ghost@example.com")
        .expect("unknown authn.begin");
    assert_eq!(
        unknown.get("state").and_then(Value::as_str),
        begun.get("state").and_then(Value::as_str)
    );
    let unknown_transaction = expect_str(&unknown, "transaction_id");
    let attempt = client
        .authn_verify_password(&unknown_transaction, password)
        .expect("unknown attempt");
    assert_eq!(attempt.get("attempts_left").and_then(Value::as_f64), Some(4.0));
    assert!(attempt.get("assurance").is_none());

    // Unknown records return stable codes rather than transport failures.
    let missing = client.tenant_get_by_name("missing").expect_err("missing tenant");
    assert_eq!(missing.code(), Some("tenant_not_found"));
    match client.request("identity.unknown", sesame::object(&[])) {
        Err(Error::Protocol(error)) => assert_eq!(error.code, "operation_not_found"),
        other => panic!("unknown operation = {other:?}"),
    }

    client.close().expect("close");
    let _ = fs::remove_dir_all(&workspace);
}
